### ISSUE TITLE
Unstripped flatpak plugins are too big.

### ISSUE THREAD

**Issue by leamas** [jongough/testplugin_pi#106]
The flatpak plugin needs to be stripped, see https://github.com/OpenCPN/OpenCPN/issues/2029. The upstream bug contains examples and a proposed patch

**Comment by rgleason** [comment]
I believe this is fixed. Try the Beta Catalog which has Testplugin, Watchdog and Pypilot..  Watchdog and Testplugin should be smaller for flatpak.  Testplugin needs ODraw. Pypilot might not work.

**Comment by leamas** [comment]
Doesn't seem so. This bug is about testplugin_pi, and I cannot see anything here related to this bug. In fact, here are no commits since end of July.

That said, that you have working code is great and means that we eventually will be able to sort this out. It's easy to check if it works, if it does the size of the flatpak plugin is roughly the same as the ubuntu one. If it doesn't work, the flatpak plugin is much larger, 8-10 times the size of the ubuntu plugin. 

Of course, the sizes refers to the library located under lib/opencpn in the tarball.

**Comment by rgleason** [comment]
This one is 338kb 
https://cloudsmith.io/~opencpn/repos/testplugin-beta/packages/detail/raw/testplugin_pi-1.0.138.0-flatpak-x86_64-18.08-flatpak-tarball/1.0.138.0+2050.ab8fdc4/

from this repository
https://github.com/rgleason/testplugin_pi

Does that look like it has been stripped?

**Comment by leamas** [comment]
yes, it's stripped.

**Comment by rgleason** [comment]
Can you try this plugin in flatpak?  
It should be available on the beta channel along with pypilot and watchdog
I would be nice know if these are working.

But Dave needs to merge first
https://github.com/OpenCPN/plugins/pull/250

**Comment by rgleason** [comment]
"yes, it's stripped."  Good. Thanks.

**Comment by rgleason** [comment]
Dave merged the PR to plugins  beta catalog.
To test these PI you would need to add  to opencpn.ini or opencpn.cong
```
[PlugIns]
catalogexpert=1

```
or download tarballs and install using Import Plugins (as you know, but I forgot)
Cloudsmith source for the plugins https://cloudsmith.io/~opencpn/repos/
https://cloudsmith.io/~opencpn/repos/pypilot-beta/packages/?q=0.22.4.0*metadata
https://cloudsmith.io/~opencpn/repos/testplugin-prod/packages/?q=1.0.137.0*metadata
https://cloudsmith.io/~opencpn/repos/watchdog-prod/packages/?q=2.4.22.0*metadata

**Comment by leamas** [comment]
@rgleason: Now, that there is a fix available for this issue is great, and that it's in the catalog is also fine. However, to be able to close this bug I suggest that you make a PR against this (Jon's) repo with your fix.

That is, assuming that testplugin_pi is maintained by Jon (?).

**Comment by rgleason** [comment]
I was waiting to get the other fixes done.

**Comment by leamas** [comment]
Nope, no good.  That is more commits than I can count. As  a project maintainer I would never merge that thing. 170 commits(!), and no description... It just don't fly, and plugin owners does the right thing  if refusing to merge. 

As things are, you have effectively forked Jon's project. You would need to discuss with Jon about your future plans: merging your different codebases or not. Until this is cleared, I'm assuming that you maintain the currently used fork of testplugin. 

Are the other PRs pi owners doesn't want to merge similar?

**Comment by rgleason** [comment]
Probably

**Comment by rgleason** [comment]
Alec,  How do I get rid of those earlier commits?  Rebase or something?
There are some things I want to keep, however.

**Comment by leamas** [comment]
Yes, you need to rebase. Now, since you havn't done this for a long time there are multiple conflicts, and you need to resolve these. The first you need to do (really!) is to find some tutorials out there how to make a git rebase. After that do something like (in your own repo)

    $  git checkout -b pr     (leave the master branch as-is as backup)
    $ git remote add upstream https://github.com/jongough/testplugin_pi.git
    $ git remote update upstream
    $ git rebase upstream/master  

The last command will stop at many commits with conflicts you have to resolve manually (you need to take that tutorial to understand this). Work through them, and you will have a history of about 100 commits which is better, but still way too much.

After rebasing, I'd recommend that you split your too long branch to manageable pieces, say about 10 commits. Make a  PR for the first 10 commits., this time with a description of what this PR does.  If the description looks sound, and the commits are readable Jon might consider to review the PR. 

Be aware that he might have remarks, and that you might have to rewrite the PR. This is the normal way to cooperate. 

Next time, be sure to make smaller PR as soon as you have done something, it  gets worse over time. Collecting things to a single PR is the wrong idea.

Also, as long as a PR isn't merged one need to rebase it  when the upstream repo changes. It's usually clean (no manual work) or with small conflicts which are easy to resolve.

**Comment by rgleason** [comment]
Thanks Leamas for the guidance. I have done this several times now, with several complete failures and one success, because I may make the wrong choices during resolution of conflicts.  I'll find a tutorial again and give it another try.

After rebasing, how do I split the long branch into 10 commits?  Is this interactive rebasing too?

**Comment by leamas** [comment]
$ git checkout pr   (assuming this is the complete branch)
    $ git checkout -b pr1  (new branch...)
    $ git reset --hard   deadbee   (the last commit to include)
    $ git push origin pr1    (and make a PR)

One need to be careful while merging. When I'm unsure, I rebuild the project after handling the conflicts but before committing. It helps. You could always start from the beginning again, something like

    $ git checkout master
    $ git branch -D pr    (drop the failed rabase attempt)
    $ git checkout -b pr   (new pr branch which is a copy of master, for a new rebase attempt)

**Comment by rgleason** [comment]
Thank you, I will study the tutorials and try to follow this.

**Comment by leamas** [comment]
You might come to a point that it isn't possible to rebase your branch since it has diverged too much from Jon's upstream. This would basically be to acknowledge that Jon is right when refusing to merge your PR, just trying to adjust his branch with patches and other input from your branch,

This would also mean that you have made an independent fork of testplugin which cannot be merged with the old upstream. The escape would then be to basically drop your your branch, ans start a new one based on upstream, like: 

    $ git branch -m master backup/master
    $ git checkout -b master upstream/master

Then, try to cherry-pick the important patches from backup/master branch starting with the oldest. This will also give conflicts, but perhaps easier to handle. This would also force you to think about the changes you have made, and why they are important.

Yes, this is painful. Forks *are* painful, there is no way to avoid that. That's we normally avoid doing forks, keeping out issue branches rebased on upstream.

**Comment by leamas** [comment]
BTW: *issue branches"... A more organized workflow from your side, similar to how I work against OpenCPN/OpenCPN:

   - When you discover a problem, file a bug (here, against testplugin).
   - Checkout a new branch having the bug number in the name: `git checkout -b bugfix-12 upstream/master`
   - Fix the problem, typically in one commit, perhaps a few. 
   - Make a PR from the issue branch. 
   - In the PR,  add a _closes: #12_ in the description. This will close your bug when the PR is merged.
   - If required (in a hurry, Jon is MIA...): make a build from the issue branch and publish it in catalog

When you see a new problem: rinse and repeat. This is the way to avoid your current situation.

**Comment by leamas** [comment]
As an example, this is how you could address the very issue in this bug:

    $ git remote rename upstream jon           (avoiding some conflicts)
    $ git remote update jon
    $ git checkout -b flatpak-106 jon/master  (start a new issue branch)
    $ git cherry-pick 1e547ff5497d   (your commit with binary: true stanza in org.opencpn.OpenCPN.yaml.in)
    $ git status  (conflicts in CMakeLists and add/rm  ci/circleci-build-android-armhf.sh)
    $ git checkout jon/master CMakeLists.txt (this commit shouldn't change this one)
    $ git rm  ci/circleci-build-android-armhf.sh  (this commit shouldn't add this one)
    $ git status (all is fine, only cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in  is changed)
    $ git commit -am "flatpak: Strip library debug symbols (#106)."
    $ git show HEAD  
    

Note that this commit also adds -DFLATPAK. Also, here is trailing whitespace after strip: true. 
Edit cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in, fix trailing whitespace

    $ git commit -a --amend  (Adjust subject to include also -DFLATPAK)
    $ git show HEAD

    commit 444f82d739cb027a60ffb2bd5c34e10b4b35aeaa (HEAD -> flatpak-106)
    Author: Rick Gleason <fcgleason@tgp-architects.com>
    Date:   Sat Sep 12 16:49:29 2020 -0400

    flatpak: Strip library debug symbols, add -DFLATPAK (#106).
    
    Besides the #106 problem, also make sure that the proper -DFLATPAK compilation
    flag is present.

    Closes: #106

    diff --git a/cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in b/cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in
    index 188db32..987266b 100644
    --- a/cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in
    +++ b/cmake/in-files/org.opencpn.OpenCPN.Plugin.yaml.in
    @@ -15,7 +15,8 @@ modules:
         - -DOCPN_FLATPAK_BUILD=ON
         build-options:
             cflags: -fPIC
    -        cxxflags: -fPIC
    +        cxxflags: -fPIC -DFLATPAK
    +        strip: true
         sources:
            - type: git
              url: git://${GIT_REPOSITORY_SERVER}/${GIT_REPOSITORY}

This is the first iteration. Now, check and fix whatever problems still around: edit files, git commit -a --amend

Eventually  all is fine. Push the flatpak-106-branch and make a PR from it. This is a PR that I happily would review and eventually accept as a project maintainer.

**Comment by rgleason** [comment]
Yikes!  Alec. Thank you. 

I see and understand the different techniques now. I think I also understand most of the git commands, and the two different ways of doing it:  fixing your own branch, or making a new branch based on Jon's repos and cherry pick a given commit.  I hope to become skilled like you at this.

The techniques using Issues and tagging commits with the issue and then closing is new to me, and I really really appreciate the big picture  "ISSUE BRANCHES"  comment above and the Organized workflow suggestions.  I have made a Text file of all of this and will annotate it as I go.

**Comment by leamas** [comment]
> and the two different ways of doing it: fixing your own branch, or making a new branch based on Jon's repos [...]

To be frank, I don't really consider "fixing you own branch" as a viable alternative. Open-source development is about cooperation, and cooperation requires a common repo/branch. Here it should IMHO be jongough/master.

That said, you could of course have your own master with some extra commits. However, to make this work you need to  rebase your own branch on jon's as soon as jon commits anything. As long as it's based on the upstream (i. e., jon) all is ok. If not, it's a pain.

Another thing is that you now have some "own branches" which cannot be rebased i. e., forks.  In the long run, you need to get away from this situation. This is done by dropping the current branch and make a fresh start based on current upstream., and keeping it in sync using rebase.

However, maintaining forks wont fly, for sure. Neither will long PRs of too many commits.

Just my 5 öre ...

PS: You should be more careful when creating commits. For example the 'strip: true' commit also contained changes to other files besides the flatpak yaml one. It would have been better to split this to two separate commits, one for the yaml file one for the CMakelists changes. This makes the commits easier to read, and much easier to handle when doing merge or rebase. If it was just the yaml file change, the cherry-pick would work without any conflicts at all...

**Comment by rgleason** [comment]
Alec wrote: If it was just the yaml file change, the cherry-pick would work without any conflicts at all...
 - Understood. 

Alec wrote: However, to make this work you need to rebase your own branch on jon's as soon as jon commits anything. As long as it's based on the upstream (i. e., jon) all is ok. If not, it's a pain.
 - Yes, I am doing this now, I am starting to work on celestial_navigation again by making a fresh start from Sean's master. I'll get back to testplugin too.

**Comment by rgleason** [comment]
https://github.com/jongough/testplugin_pi/pull/114

**Comment by jongough** [comment]
Flatpak should be much smaller now. Change made in last day and applied to Master branch

**Comment by leamas** [comment]
Yes, a somewhat hidden part of e17e1449 :) Closing
